src/Compiler/CompilerPasses.cpp

```diff
@@ -121,17 +122,33 @@ void addONNXToKrnlPasses(mlir::PassManager &pm, int optLevel, bool enableCSE,
   if (enableInstrumentONNXSignature)
     pm.addNestedPass<func::FuncOp>(
         onnx_mlir::createInstrumentONNXSignaturePass());
+  if (enableLinalg) {
+    pm.addPass(onnx_mlir::createLowerONNXToLinalgPass());
+    // Linalg bufferization can be before or after LowerToKrnlPass
+    // However, sice bufferization::AllocTensorOp is in LowerToKrnl temporarily,
+    // these passes have to be called before LowerToKrnlPass
+    pm.addNestedPass<func::FuncOp>(createLinalgBufferizePass());
+    // Convert tensor.EmptyOp to bufferization.alloc_tensor
+    pm.addNestedPass<func::FuncOp>(
+        bufferization::createEmptyTensorToAllocTensorPass());
+  }
   pm.addPass(onnx_mlir::createLowerToKrnlPass(optLevel, enableParallel));
   // An additional pass of canonicalization is helpful because lowering
   // from ONNX dialect to Standard dialect exposes additional canonicalization
   // opportunities.
+
+  // For Linalg and Krnl mixed IR:
+  // Canonicalization pass will clean up bufferization::to_tensor and to_memref
```

Comment:
We needed to add a new pass to accomplish eliminate type conversions, and looking at your IR output it seems you still have bufferization::to_memref showing up.

Strictly speaking, I don't know if such a pass should be necessary, and in your case it may be because you aren't doing one shot bufferization, but if that doesn't clean things up, you might want to look at adding a pass for cleaning up type conversions.

Reply:
In this particular test case, the to_tensor/to_memref are cleaned up by canonicalization pass. Since ONNX has data types other than tensor, such as Sequence and Map type, I think we may have to convert to_tensor/to_memref to UnrealizedConversionCastOp so that they can be cancelled out by canonicalization. The cast Op can handle any type.